### PR TITLE
Binheap

### PR DESCRIPTION
**Pull request by nadiabahrami**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,3 +28,10 @@ Anothe Node-chain setup composed with double linked list but with the following
     - append(val)
     - appendleft(val)
 
+
+
+## Binary_heap.py
+
+A heap data-structure defined by wikipedia, found here: https://en.wikipedia.org/wiki/Binary_heap
```
Comment (joelstanner) [inline README.md:35]: At some point it would be good to write your own description of this data structure to demonstrate your understanding of it to potential employers.

File: src/binary_heap.py [hunk 2]
```diff
@@ -3,52 +3,45 @@
...
+                for i, value in enumerate(self.heap):
+                    if i == 0:
+                        continue
+                    if biggest < value:
+                        biggest = value
+                    try:
+                        if value < self.heap[i * 2]:
+                            self.heap[i], self.heap[2 * i] = self.heap[i * 2], value
+                        if value < self.heap[(i * 2) + 1]:
+                            self.heap[i], self.heap[(2 * i) + 1] = self.heap[(i * 2) + 1], value
```
Comment (joelstanner) [inline src/binary_heap.py:28]: For this class, we want to adhere to the PEP8 80 character limit (since we're mostly on laptops). You could write these like this:
...
removed
...

Ok not quite like that --- but close.

See python's [implicit line joining](https://google.github.io/styleguide/pyguide.html?showone=Line_length#Line_length).

File: src/binary_heap.py [hunk 3]
```diff
@@ -3,52 +3,45 @@
...
-
-    def push(self, val):
-        """Pushes a value to end of self.heap while maintaining structure."""
-        self.heap.append(val)
-        self.sift(0, self.length - 1)
+    """Define a  max BinaryHeap object."""
+
+    def __init__(self):
+        """Initialize heap object."""
+        self.heap = []
```
Comment (joelstanner) [inline src/binary_heap.py:10]: From the assignment description: "The constructor for your heap should default to creating an empty heap, but allow for creating a populated given an iterable as an input." This constructor does not allow for passing input.

File: src/binary_heap.py [hunk 4]
```diff
@@ -3,52 +3,45 @@
...
+    def restructure(self):
+        """Maintain structure of heap to max."""
+        if len(self.heap) > 1:  # Solves 0 and 1 case THIS IS GOOD'
+            biggest = self.heap[-1]
+
+            while biggest != self.heap[0]:
+
+                for i, value in enumerate(self.heap):
+                    if i == 0:
+                        continue
```
Comment (joelstanner) [inline src/binary_heap.py:21]: An interesting approach to skipping index 0. `enumerate()` also accepts a `start` argument, so you could also do 

``` python
enumerate(self.heap, start=1)
```

File: src/binary_heap.py [hunk 5]
```diff
@@ -3,52 +3,45 @@
...
-        """Pushes a value to end of self.heap while maintaining structure."""
-        self.heap.append(val)
-        self.sift(0, self.length - 1)
+    """Define a  max BinaryHeap object."""
+
+    def __init__(self):
+        """Initialize heap object."""
+        self.heap = []
+
+    def restructure(self):
```
Comment (joelstanner) [inline src/binary_heap.py:12]: An end user won't ever need to call this function (because it is called automatically from your `push()` and `pop()`. The pythonic way of designating a semi-private method is prepending it with an underscore, so: `_restructure(self)`.

File: src/test_binary_heap.py [hunk 6]
```diff
@@ -3,23 +3,47 @@
...
+    binary_heap_object.push(1)
+    assert binary_heap_object.heap == [1]
+
+
+def test_binary_heap_object_2(binary_heap_object):
+    """Test a single push."""
+    binary_heap_object.push(1)
+    binary_heap_object.push(2)
+    assert binary_heap_object.heap == [2, 1]
+
```
Comment (joelstanner) [inline src/test_binary_heap.py:34]: It would be good to also test edge cases for `pop()` with 0, 1, and 2 items in the heap.
